src/Microsoft.Extensions.Configuration.AzureAppConfiguration/ConfigurationClientManager.cs

```diff
@@ -118,8 +223,137 @@ public Uri GetEndpointForClient(ConfigurationClient client)
             }
 
             ConfigurationClientWrapper currentClient = _clients.FirstOrDefault(c => c.Client == client);
-            
+
+            if (_dynamicClients != null && currentClient == null)
+            {
+                currentClient = _dynamicClients.FirstOrDefault(c => c.Client == client);
+            }
+
             return currentClient?.Endpoint;
         }
+
+        public void SetLogger(Logger logger)
+        {
+            if (logger == null)
+            {
+                throw new ArgumentNullException(nameof(logger));
+            }
+
+            _logger = logger;
+        }
+
+        private async Task OptimisticRefreshFallbackClients(DateTimeOffset now, CancellationToken cancellationToken)
```

Comment:
This is a task that may often do no work (starts with an `if` statement)

I usually stay away from methods that are surrounded by an `if` statement. It's a good indicator that the caller should be checking the condition and not invoking the method if the condition is not met.

That is doubly true if the cost is an unawaited task like what we have here.

Let's have the caller check whether refresh is needed. That way we don't spin up a task unnecessarily.

Also, with the change I mentioned in my previous comment, we'll need a name change. I'm imagining a pattern like

```
if (IsFallbackClientDiscoveryDue())
{
    _ = TriggerFallbackClientDiscovery(cancellationToken)
        .SwallowCancellation();
}
```